libpromises/evalfunction.c

```diff
@@ -8336,6 +8336,116 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
 
 /*********************************************************************/
 
+static FnCallResult FnCallFindfilesUp(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+
+    if (finalargs == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires path as first argument",
+            fp->name);
+        return FnFailure();
+    }
+
+    char path[PATH_MAX];
+    size_t copied = strlcpy(path, RlistScalarValue(finalargs), sizeof(path));
+    if (copied >= sizeof(path))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, path is too long (%zu > %zu)",
+            fp->name, copied, sizeof(path));
+        return FnFailure();
+    }
+
+    if (!IsAbsoluteFileName(path))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, not an absolute path '%s'",
+            fp->name, path);
+        return FnFailure();
+    }
+
+    if (!IsDir(path))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, path '%s' is not a directory",
+            fp->name, path);
+        return FnFailure();
+    }
+
+    DeleteRedundantSlashes(path);
+
+    size_t len = strlen(path);
+    if (path[len - 1] == FILE_SEPARATOR)
+    {
+        path[len - 1] = '\0';
+    }
+
+    if (finalargs->next == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires glob as second argument",
+            fp->name);
+        return FnFailure();
+    }
+
+    char glob[PATH_MAX];
+    copied = strlcpy(glob, RlistScalarValue(finalargs->next), sizeof(glob));
+    if (copied >= sizeof(glob))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, glob is too long (%zu > %zu)",
+            fp->name, copied, sizeof(glob));
+        return FnFailure();
+    }
+
+    if (IsAbsoluteFileName(glob))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Function %s, glob '%s' cannot be an aboslute path", fp->name,
+            glob);
+        return FnFailure();
+    }
+
+    DeleteRedundantSlashes(glob);
+
+    /* level defaults to inf */
+    long level = IntFromString("inf");
+    if (finalargs->next->next != NULL)
+    {
+        level = IntFromString(RlistScalarValue(finalargs->next->next));
+    }
+
+    JsonElement *json = JsonArrayCreate(1);
+
+    while (level-- >= 0)
+    {
+        char filepath[PATH_MAX];
+        copied = strlcpy(filepath, path, sizeof(filepath));
+        assert(copied < sizeof(path));
+        if (JoinPaths(filepath, sizeof(filepath), glob) == NULL)
+        {
+            Log(LOG_LEVEL_ERR,
+                "Function %s, failed to join glob '%s' and path '%s'",
```

Comment:
It appends the glob to the path, right? I think the log message should say that or *join path and glob*.